proto/backup.proto

```diff
@@ -185,8 +187,27 @@ message GCS {
     // You can get one from https://console.cloud.google.com/apis/credentials, and
     // copy the content, set it as string here.
     string credentials_blob = 6;
+    Bucket bucket_info = 7;
 }
 
+// S3 and GCS can be transitioned to use this
+message Bucket {
+    string endpoint = 1; // use a different endpoint (host + url)
+    string host = 2; // use a different host
```

Comment:
why we need a `host` field

Reply:
For GCS it is easier to give a host `localhost:4443` rather than having to know the full url of `localhost:4443/storage/v1`

Reply:
this can be manipulated in the user interface, no need to be represented the gRPC protocol.

Reply:
I don't know what you mean. Do you mean to give just a host as the endpoint and have TiKV figure out the intent?

Reply:
let's say BR provides both `--gcs.endpoint` and `--gcs.host` parameters. then BR will send

```go
endpoint := gcs.endpoint
if endpoint == "" {
    endpoint = gcs.host + "/storage/v1"
}
```

to TiKV.

But for GCS... I thought the endpoint URL is *still* `http://localhost:4443`? The GCS library should append the `/storage/v1` part itself when constructing the URL.

Reply:
The endpoint url needs /storage/v1, at least for fake-gcs-server. TiKV does not automatically add it. I specifically tried this and then wrote unit tests to verify the current behavior. I didn't want to make a breaking change in behavior so I created a new field.

I agree that the code you added to BR makes sense given what I know. My concerns are:
* Is there some use case where storage/v1 should not be added to the url? What happens when TiKV switches to storage/v2 ?
* external_storage can be used for more than just backup in the future. TiKV or PD could be storing potentially any of their data to external_storage. Physical backup could be run automatically from TiKV without an external tool.

Reply:
Ok. Well the situation is not good...

* Currently GCS + custom endpoint is not well tested on BR (pingcap/br#258), so I hope we are allowed to break compatibility for this experimentally if a more meaningful choice can be taken 🙃.

* > Is there some use case where storage/v1 should not be added to the url? 

  Maybe a custom GCS service won't bother to attach the storage/v1 part, dunno. If there is a storage/v2 then it makes more sense to provide the "host" version of endpoint only.

* > TiKV or PD could be storing potentially any of their data to external_storage. 

  TiKV and PD and do the host → endpoint URL normalization too.

Reply:
I tested the GCS endpoint flag against gcs-fake-server and was able to get it to work. The command was written down here: https://github.com/tikv/tikv/pull/9706
I also tested setting the host with the scli example.

If we look at the [GCS storage API](https://cloud.google.com/storage/docs/request-endpoints), there are multiple url endpoints.

* https://storage.googleapis.com/storage/v1/PATH_TO_RESOURCE
* https://storage.googleapis.com/upload/storage/v1/b/BUCKET_NAME/o
* https://storage.googleapis.com/batch/storage/v1/PATH_TO_RESOURCE
* https://storage.googleapis.com/download/storage/v1/b/BUCKET_NAME/o/OBJECT_NAME?alt=media

The TiKV code has 2 of them now. To me it doesn't make sense to be giving one of the urls as the endpoint. Actually I think this code will already fail if a user gives storage/v1 but the API needs to use a different url.
What makes sense to me is to that the endpoint is the common subset of urls. Effectively the host flag does this now. I didn't want to alter the existing behavior of the endpoint field and break backwards compatibility.

Reply:
[This test case in my PR](https://github.com/tikv/tikv/pull/9706/files#diff-60ec9b1be5e8a5a1732ccbab6e3c6aa7d1643f059bd1347edca5e62d236eb687R510) demonstrates that the system of overriding the full endpoint url is broken. That is it always changes the full url even though there are 2 different urls that should be in use.

That PR now has more merge conflicts to resolve because I am waiting on this PR to be merged. Can we come to a resolution on this conversation?

Reply:
I'll send this to the team for discussion tomorrow. My preference:

* `--gcs.endpoint` doesn't need to care about backward compatibility since it's not documented at all
* the endpoint should have the same behavior as https://pkg.go.dev/google.golang.org/api/option#WithEndpoint. AFAIK both of the popular Rust GCS crates `tame_gcs` and `cloud-storage` hardcoded the endpoint URL 😢
* If `--gcs.host` and `--gcs.endpoint` are different by default, the former should be resolved to the endpoint URL at the client library level. I prefer not to have both the host and endpoint fields in BucketInfo, since it is unnecessary for S3. Client-side normalization is also how we handle `--s3.provider` vs "force-path-style".

Reply:
For `S3` I don't have the host field, only for `Gcs` and CloudDynamic`. We could remove host from `Bucket` used by `CloudDynamic` as well. 

I would be happy to change the behavior of `gcs.endpoint`

Note that it is possible (although obviously unwieldy) when using the client libraries to create a client per url. I suspect that they are both poorly designed and almost nobody tries to use that option.

Reply:
We've discussed about the compatibility issue for BR, and the conclusion is that:

"BR GCS support is still not released as stable. Therefore, we accept incompatible changes."

----

> Note that it is possible (although obviously unwieldy) when using the client libraries to create a client per url

yeah we support both the parameter style `-s 'gcs://foo/bar' --gcs.endpoint https://1.2.3.4:5678` and URL style `-s 'gcs://foo/bar?endpoint=https://1.2.3.4:5678'`. The same for `host` if it exists.

Reply:
So we should change the behavior of `endpoint` for GCS to not have /storage/v1? Then we don't need the `host` field.

Reply:
> So we should change the behavior of `endpoint` for GCS to not have /storage/v1

SGTM.

Reply:
okay, that change is on this PR now and I will make the change on the TiKV PR as well.